File: demo/zig/src/auth/handler.zig
const std = @import("std");
const c = @cImport({
    @cInclude("sqlite3.h");
});

const password = @import("password.zig");
const email = @import("email.zig");
const service = @import("service.zig");
const errors = @import("errors.zig");

const Server = std.http.Server;

pub fn dispatch(
    allocator: std.mem.Allocator,
    db: *c.sqlite3,
    res: *Server.Response,
) !void {
    const req = res.request;
    const target = req.target;
    const method = req.method;

    const path = if (std.mem.indexOfScalar(u8, target, '?')) |qp| target[0..qp] else target;

    if (std.mem.eql(u8, path, "/api/auth/signup") and method == .POST) {
        try handleSignUp(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/signin") and method == .POST) {
        try handleSignIn(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/signout") and method == .POST) {
        try handleSignOut(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/session") and method == .GET) {
        try handleSession(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/verify-email") and (method == .GET or method == .POST)) {
        try handleVerifyEmail(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/forgot-password") and method == .POST) {
        try handleForgotPassword(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/api/auth/reset-password") and method == .POST) {
        try handleResetPassword(allocator, db, res);
    } else if (std.mem.eql(u8, path, "/health") and method == .GET) {
        try handleHealth(res);
    } else {
        res.status = .not_found;
        try respondJson(res, "{\"error\":\"not found\"}");
    }
}

fn respondJson(res: *Server.Response, json: []const u8) !void {
    res.transfer_encoding = .{ .content_length = json.len };
    try res.headers.append("content-type", "application/json");
    try res.do();
    try res.writeAll(json);
    try res.finish();
}

fn respondJsonStatus(res: *Server.Response, status: Server.Response.Status, json: []const u8) !void {
    res.status = status;
    try respondJson(res, json);
}

fn respondJsonWithCookies(res: *Server.Response, status: Server.Response.Status, json: []const u8, cookies: []const u8) !void {
    res.status = status;
    res.transfer_encoding = .{ .content_length = json.len };
    try res.headers.append("content-type", "application/json");
    if (cookies.len > 0) {
        try res.headers.append("set-cookie", cookies);
    }
    try res.do();
    try res.writeAll(json);
    try res.finish();
}

fn getSessionTokenFromRequest(res: *Server.Response) ?[]const u8 {
    const header_value = res.request.headers.get("cookie") orelse return null;
    var iter = std.mem.splitSequence(u8, header_value, ";");
    while (iter.next()) |part| {
        const trimmed = std.mem.trim(u8, part, " ");
        if (std.mem.startsWith(u8, trimmed, "session=")) {
            return trimmed["session=".len..];
        }
    }
    return null;
}

fn readRequestBody(res: *Server.Response, allocator: std.mem.Allocator) ![]u8 {
    var buf = std.ArrayList(u8).init(allocator);
    defer buf.deinit();

    const reader = res.request.reader();
    try reader.readAllArrayList(&buf, 1_000_000);

    return try buf.toOwnedSlice();
}

fn getQueryParam(res: *Server.Response, name: []const u8) ?[]const u8 {
    const target = res.request.target;
    const qm = std.mem.indexOfScalar(u8, target, '?') orelse return null;
    const query = target[qm + 1 ..];

    var iter = std.mem.splitSequence(u8, query, "&");
    while (iter.next()) |pair| {
        if (std.mem.indexOfScalar(u8, pair, '=')) |eq| {
            const key = pair[0..eq];
            if (std.mem.eql(u8, key, name)) {
                const value = pair[eq + 1 ..];
                if (value.len == 0) return "";
                return value;
            }
        }
    }
    return null;
}

fn handleHealth(res: *Server.Response) !void {
    try respondJson(res, "{\"status\":\"ok\"}");
}

fn handleSignUp(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    const body = try readRequestBody(res, allocator);
    defer allocator.free(body);

    const parsed = std.json.parseFromSlice(
        std.json.Value,
        allocator,
        body,
        .{},
    ) catch {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"invalid json\"}");
    };
    defer parsed.deinit();

    const obj = parsed.value.object;
    const email_str = obj.get("email") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    };
    const password_str = obj.get("password") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password is required\"}");
    };

    if (email_str != .string or email_str.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    }
    if (password_str != .string or password_str.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password is required\"}");
    }

    const user = service.createUser(allocator, db, email_str.string, password_str.string) catch |err| {
        switch (err) {
            errors.AuthError.EmailTaken => return respondJsonStatus(res, .@"conflict", "{\"error\":\"could not create that account\"}"),
            else => {
                std.log.err("signup: {s}", .{@errorName(err)});
                return respondJsonStatus(res, .internal_server_error, "{\"error\":\"could not create that account\"}");
            },
        }
    };

    service.createVerificationToken(allocator, db, user.email) catch {};
    email.sendVerificationEmail(allocator, user.email, user.name, "");

    try respondJsonStatus(res, .created, "{\"message\":\"Check your email to verify your account before signing in.\"}");
}

fn handleSignIn(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    const body = try readRequestBody(res, allocator);
    defer allocator.free(body);

    const parsed = std.json.parseFromSlice(
        std.json.Value,
        allocator,
        body,
        .{},
    ) catch {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"invalid json\"}");
    };
    defer parsed.deinit();

    const obj = parsed.value.object;
    const email_str = obj.get("email") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    };
    const password_str = obj.get("password") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password is required\"}");
    };

    if (email_str != .string or email_str.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    }
    if (password_str != .string or password_str.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password is required\"}");
    }

    const result = service.getUserByEmail(allocator, db, email_str.string) catch {
        return respondJsonStatus(res, .unauthorized, "{\"error\":\"invalid email or password\"}");
    };

    if (!password.checkPassword(password_str.string, result.password_hash)) {
        defer allocator.free(result.password_hash);
        return respondJsonStatus(res, .unauthorized, "{\"error\":\"invalid email or password\"}");
    }
    defer allocator.free(result.password_hash);

    if (!result.user.email_verified) {
        const vt = service.createVerificationToken(allocator, db, result.user.email) catch null;
        if (vt) |token| {
            defer allocator.free(token);
            email.sendVerificationEmail(allocator, result.user.email, result.user.name, token);
        }
        return respondJsonStatus(res, .forbidden, "{\"error\":\"please verify your email address before signing in. we sent a new verification link\"}");
    }

    const session = service.createSession(allocator, db, result.user.id) catch {
        return respondJsonStatus(res, .internal_server_error, "{\"error\":\"could not sign in\"}");
    };

    const cookie = std.fmt.allocPrint(allocator, "session={s}; Path=/; HttpOnly; SameSite=Lax; Max-Age={d}", .{
        session.token,
        30 * 24 * 60 * 60,
    }) catch {
        return respondJsonStatus(res, .internal_server_error, "{\"error\":\"could not sign in\"}");
    };
    defer allocator.free(cookie);

    const user_json = std.fmt.allocPrint(allocator,
        \\{{"user":{{"id":"{s}","email":"{s}","name":"{s}","emailVerified":{s},"createdAt":"{s}"}},"redirectTo":"/"}}
    , .{
        result.user.id,
        result.user.email,
        result.user.name,
        if (result.user.email_verified) "true" else "false",
        result.user.created_at,
    }) catch {
        return respondJsonStatus(res, .internal_server_error, "{\"error\":\"could not sign in\"}");
    };
    defer allocator.free(user_json);

    try respondJsonWithCookies(res, .ok, user_json, cookie);
}

fn handleSignOut(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    if (getSessionTokenFromRequest(res)) |token| {
        service.deleteSession(db, token) catch {};
    }

    const clear_cookie = "session=; Path=/; HttpOnly; SameSite=Lax; Max-Age=-1";
    _ = allocator;

    try respondJsonWithCookies(res, .ok, "{\"message\":\"signed out\"}", clear_cookie);
}

fn handleSession(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    const token = getSessionTokenFromRequest(res) orelse {
        return respondJson(res, "{\"user\":null}");
    };

    const result = service.getSession(allocator, db, token) catch null;
    if (result) |r| {
        const user_json = std.fmt.allocPrint(allocator,
            \\{{"user":{{"id":"{s}","email":"{s}","name":"{s}","emailVerified":{s},"createdAt":"{s}"}}}}
        , .{
            r.user.id,
            r.user.email,
            r.user.name,
            if (r.user.email_verified) "true" else "false",
            r.user.created_at,
        }) catch {
            return respondJson(res, "{\"user\":null}");
        };
        defer allocator.free(user_json);
        return respondJson(res, user_json);
    }

    return respondJson(res, "{\"user\":null}");
}

fn handleVerifyEmail(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    var token_val: ?[]const u8 = getQueryParam(res, "token");

    if (token_val == null and res.request.method == .POST) {
        const body = try readRequestBody(res, allocator);
        defer allocator.free(body);

        const parsed = std.json.parseFromSlice(
            std.json.Value,
            allocator,
            body,
            .{},
        ) catch null;
        if (parsed) |p| {
            defer p.deinit();
            if (p.value.object.get("token")) |t| {
                if (t == .string and t.string.len > 0) {
                    token_val = t.string;
                }
            }
        }
    }

    const token = token_val orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"token is required\"}");
    };

    const redirect_url = std.fmt.allocPrint(allocator, "/signin?verified=1", .{}) catch {
        return respondJsonStatus(res, .internal_server_error, "{\"error\":\"internal error\"}");
    };
    defer allocator.free(redirect_url);

    service.verifyEmail(db, token) catch |err| {
        switch (err) {
            errors.AuthError.InvalidToken, errors.AuthError.TokenExpired => {
                return respondRedirect(res, "/signin?error=This+verification+link+is+invalid+or+expired");
            },
            else => {
                std.log.err("verify email: {s}", .{@errorName(err)});
                return respondRedirect(res, "/signin?error=Could+not+verify+email");
            },
        }
    };

    return respondRedirect(res, redirect_url);
}

fn handleForgotPassword(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    const body = try readRequestBody(res, allocator);
    defer allocator.free(body);

    const parsed = std.json.parseFromSlice(
        std.json.Value,
        allocator,
        body,
        .{},
    ) catch {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"invalid json\"}");
    };
    defer parsed.deinit();

    const obj = parsed.value.object;
    const email_str = obj.get("email") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    };

    if (email_str != .string or email_str.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"email is required\"}");
    }

    const message = "{\"message\":\"if that email exists, a password reset link has been sent\"}";
    const exists = service.userExists(db, email_str.string) catch {
        return respondJson(res, message);
    };

    if (!exists) {
        return respondJson(res, message);
    }

    const vt_token = service.createVerificationToken(allocator, db, email_str.string) catch null;
    if (vt_token) |token| {
        defer allocator.free(token);
        email.sendPasswordResetEmail(allocator, email_str.string, email_str.string, token);
    }

    return respondJson(res, message);
}

fn handleResetPassword(allocator: std.mem.Allocator, db: *c.sqlite3, res: *Server.Response) !void {
    const body = try readRequestBody(res, allocator);
    defer allocator.free(body);

    const parsed = std.json.parseFromSlice(
        std.json.Value,
        allocator,
        body,
        .{},
    ) catch {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"invalid json\"}");
    };
    defer parsed.deinit();

    const obj = parsed.value.object;

    const token_val = obj.get("token") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"reset token is missing\"}");
    };
    if (token_val != .string or token_val.string.len == 0) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"reset token is missing\"}");
    }

    const password_val = obj.get("password") orelse {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password must be at least 8 characters\"}");
    };
    if (password_val != .string or password_val.string.len < 8) {
        return respondJsonStatus(res, .bad_request, "{\"error\":\"password must be at least 8 characters\"}");
    }

    service.resetPassword(allocator, db, token_val.string, password_val.string) catch |err| {
        switch (err) {
            errors.AuthError.InvalidToken, errors.AuthError.TokenExpired => {
                return respondJsonStatus(res, .bad_request, "{\"error\":\"that reset link is invalid or expired\"}");
            },
            else => {
                std.log.err("reset password: {s}", .{@errorName(err)});
                return respondJsonStatus(res, .internal_server_error, "{\"error\":\"could not reset password\"}");
            },
        }
    };

    return respondJson(res, "{\"message\":\"your password has been reset\"}");
}

fn respondRedirect(res: *Server.Response, location: []const u8) !void {
    res.status = .temporary_redirect;
    res.transfer_encoding = .{ .content_length = 0 };
    try res.headers.append("location", location);
    try res.do();
    try res.finish();
}

File: demo/rust/src/auth/handler.rs
use super::email::{send_password_reset_email, send_verification_email};
use super::errors::AuthError;
use super::password;
use super::service::Service;
use axum::{
    Json, Router,
    extract::{Query, State},
    http::{HeaderMap, HeaderValue, StatusCode, header},
    response::{IntoResponse, Redirect, Response},
    routing::{get, post},
};
use serde::Deserialize;
use serde_json::{Value, json};
use std::sync::{Arc, Mutex};
use rusqlite::Connection;

#[derive(Clone)]
pub struct AppState {
    pub db: Arc<Mutex<Connection>>,
}

pub fn routes(state: AppState) -> Router {
    Router::new()
        .route("/api/auth/signup", post(sign_up))
        .route("/api/auth/signin", post(sign_in))
        .route("/api/auth/signout", post(sign_out))
        .route("/api/auth/session", get(session))
        .route("/api/auth/verify-email", get(verify_email_get).post(verify_email_post))
        .route("/api/auth/forgot-password", post(forgot_password))
        .route("/api/auth/reset-password", post(reset_password))
        .route("/health", get(health))
        .with_state(state)
}

fn json_response(status: StatusCode, value: Value) -> Response {
    (status, Json(value)).into_response()
}

fn json_error(status: StatusCode, message: &str) -> Response {
    json_response(status, json!({ "error": message }))
}

fn session_token(headers: &HeaderMap) -> Option<String> {
    let cookie_header = headers.get(header::COOKIE)?.to_str().ok()?;
    cookie_header
        .split(';')
        .map(str::trim)
        .find_map(|part| part.strip_prefix("session=").map(str::trim).map(str::to_string))
}

fn set_session_cookie(headers: &mut HeaderMap, token: &str) {
    let value = format!(
        "session={token}; Path=/; HttpOnly; SameSite=Lax; Max-Age={}",
        30 * 24 * 60 * 60
    );
    if let Ok(header_value) = HeaderValue::from_str(&value) {
        headers.append(header::SET_COOKIE, header_value);
    }
}

fn clear_session_cookie(headers: &mut HeaderMap) {
    let value = "session=; Path=/; HttpOnly; SameSite=Lax; Max-Age=-1";
    if let Ok(header_value) = HeaderValue::from_str(value) {
        headers.append(header::SET_COOKIE, header_value);
    }
}

#[derive(Deserialize)]
struct SignUpRequest {
    email: String,
    password: String,
}

#[derive(Deserialize)]
struct SignInRequest {
    email: String,
    password: String,
}

#[derive(Deserialize)]
struct EmailRequest {
    email: String,
}

#[derive(Deserialize)]
struct TokenRequest {
    token: String,
}

#[derive(Deserialize)]
struct ResetPasswordRequest {
    token: String,
    password: String,
}

#[derive(Deserialize)]
struct VerifyEmailQuery {
    token: Option<String>,
}

async fn sign_up(
    State(state): State<AppState>,
    Json(mut req): Json<SignUpRequest>,
) -> Response {
    req.email = req.email.trim().to_string();
    if req.email.is_empty() || req.password.is_empty() {
        return json_error(StatusCode::BAD_REQUEST, "email and password are required");
    }

    let conn = state.db.lock().unwrap();
    let user = match Service::create_user(&conn, &req.email, &req.password) {
        Ok(user) => user,
        Err(AuthError::EmailTaken) => {
            return json_error(StatusCode::CONFLICT, "could not create that account");
        }
        Err(_) => {
            eprintln!("signup error");
            return json_error(
                StatusCode::INTERNAL_SERVER_ERROR,
                "could not create that account",
            );
        }
    };

    if let Ok(vt_token) = Service::create_verification_token(&conn, &user.email) {
        let email = user.email.clone();
        let name = user.name.clone();
        send_verification_email(&email, &name, &vt_token);
    } else {
        eprintln!("create verification token error");
    }

    json_response(
        StatusCode::CREATED,
        json!({ "message": "Check your email to verify your account before signing in." }),
    )
}

async fn sign_in(
    State(state): State<AppState>,
    Json(mut req): Json<SignInRequest>,
) -> Response {
    req.email = req.email.trim().to_string();
    if req.email.is_empty() || req.password.is_empty() {
        return json_error(StatusCode::BAD_REQUEST, "email and password are required");
    }

    let conn = state.db.lock().unwrap();
    let (user, hash) = match Service::get_user_by_email(&conn, &req.email) {
        Ok(pair) => pair,
        Err(_) => {
            return json_error(StatusCode::UNAUTHORIZED, "invalid email or password");
        }
    };

    if !password::check_password(&req.password, &hash) {
        return json_error(StatusCode::UNAUTHORIZED, "invalid email or password");
    }

    if !user.email_verified {
        if let Ok(vt_token) = Service::create_verification_token(&conn, &user.email) {
            let email = user.email.clone();
            let name = user.name.clone();
            send_verification_email(&email, &name, &vt_token);
        } else {
            eprintln!("create verification token error");
        }
        return json_error(
            StatusCode::FORBIDDEN,
            "please verify your email address before signing in. we sent a new verification link",
        );
    }

    let session = match Service::create_session(&conn, &user.id) {
        Ok(session) => session,
        Err(err) => {
            eprintln!("create session error: {err}");
            return json_error(StatusCode::INTERNAL_SERVER_ERROR, "could not sign in");
        }
    };

    let mut headers = HeaderMap::new();
    set_session_cookie(&mut headers, &session.token);
    let body = json!({ "user": user, "redirectTo": "/" });
    let mut response = json_response(StatusCode::OK, body);
    response.headers_mut().extend(headers);
    response
}

async fn sign_out(State(state): State<AppState>, headers: HeaderMap) -> Response {
    if let Some(token) = session_token(&headers) {
        let conn = state.db.lock().unwrap();
        let _ = Service::delete_session(&conn, &token);
    }

    let mut response_headers = HeaderMap::new();
    clear_session_cookie(&mut response_headers);
    let mut response = json_response(StatusCode::OK, json!({ "message": "signed out" }));
    response.headers_mut().extend(response_headers);
    response
}

async fn session(State(state): State<AppState>, headers: HeaderMap) -> Response {
    let Some(token) = session_token(&headers) else {
        return json_response(StatusCode::OK, json!({ "user": null }));
    };

    let conn = state.db.lock().unwrap();
    match Service::get_session(&conn, &token) {
        Ok(Some((_, user))) => json_response(StatusCode::OK, json!({ "user": user })),
        Ok(None) | Err(_) => json_response(StatusCode::OK, json!({ "user": null })),
    }
}

async fn verify_email_get(
    State(state): State<AppState>,
    Query(query): Query<VerifyEmailQuery>,
) -> Response {
    let Some(token) = query.token.filter(|t| !t.is_empty()) else {
        return json_error(StatusCode::BAD_REQUEST, "token is required");
    };
    verify_email_redirect(&state, &token).await
}

async fn verify_email_post(
    State(state): State<AppState>,
    Query(query): Query<VerifyEmailQuery>,
    body: Option<Json<TokenRequest>>,
) -> Response {
    let token = query.token.clone().filter(|t| !t.is_empty()).or_else(|| {
        body.as_ref()
            .map(|Json(req)| req.token.clone())
            .filter(|t| !t.is_empty())
    });

    let Some(token) = token else {
        return json_error(StatusCode::BAD_REQUEST, "token is required");
    };
    verify_email_redirect(&state, token.as_str()).await
}

async fn verify_email_redirect(state: &AppState, token: &str) -> Response {
    let conn = state.db.lock().unwrap();
    match Service::verify_email(&conn, token) {
        Ok(()) => Redirect::temporary("/signin?verified=1").into_response(),
        Err(AuthError::InvalidToken) | Err(AuthError::TokenExpired) => {
            Redirect::temporary("/signin?error=This+verification+link+is+invalid+or+expired")
                .into_response()
        }
        Err(_) => {
            eprintln!("verify email error");
            Redirect::temporary("/signin?error=Could+not+verify+email").into_response()
        }
    }
}

async fn forgot_password(
    State(state): State<AppState>,
    Json(mut req): Json<EmailRequest>,
) -> Response {
    req.email = req.email.trim().to_string();
    if req.email.is_empty() {
        return json_error(StatusCode::BAD_REQUEST, "email is required");
    }

    let message = "if that email exists, a password reset link has been sent";
    let conn = state.db.lock().unwrap();

    let exists = match Service::user_exists(&conn, &req.email) {
        Ok(exists) => exists,
        Err(err) => {
            eprintln!("user exists check error: {err}");
            return json_response(StatusCode::OK, json!({ "message": message }));
        }
    };

    if !exists {
        return json_response(StatusCode::OK, json!({ "message": message }));
    }

    match Service::create_verification_token(&conn, &req.email) {
        Ok(token) => {
            let email = req.email.clone();
            send_password_reset_email(&email, &email, &token);
        }
        Err(err) => eprintln!("create verification token error: {err}"),
    }

    json_response(StatusCode::OK, json!({ "message": message }))
}

async fn reset_password(
    State(state): State<AppState>,
    Json(req): Json<ResetPasswordRequest>,
) -> Response {
    if req.token.is_empty() {
        return json_error(StatusCode::BAD_REQUEST, "reset token is missing");
    }
    if req.password.is_empty() || req.password.len() < 8 {
        return json_error(
            StatusCode::BAD_REQUEST,
            "password must be at least 8 characters",
        );
    }

    let conn = state.db.lock().unwrap();
    match Service::reset_password(&conn, &req.token, &req.password) {
        Ok(()) => json_response(
            StatusCode::OK,
            json!({ "message": "your password has been reset" }),
        ),
        Err(AuthError::InvalidToken) | Err(AuthError::TokenExpired) => {
            json_error(StatusCode::BAD_REQUEST, "that reset link is invalid or expired")
        }
        Err(_) => {
            eprintln!("reset password error");
            json_error(StatusCode::INTERNAL_SERVER_ERROR, "could not reset password")
        }
    }
}

async fn health() -> Response {
    json_response(StatusCode::OK, json!({ "status": "ok" }))
}
